profiler/src/main/java/com/splunk/opentelemetry/profiler/snapshot/PeriodicStackTraceSampler.java

```diff
@@ -0,0 +1,234 @@
+          traceId,
+          tid -> {
+            SamplingContext context = new SamplingContext(thread, tid, spanId, System.nanoTime());
+            takeOnDemandSample(context).ifPresent(staging.get()::stage);
+            return context;
+          });
+    }
+
+    void remove(String traceId, String spanId) {
+      traceSamplingContexts.computeIfPresent(
+          traceId,
+          (t, context) -> {
+            if (spanId.equals(context.spanId)) {
+              takeOnDemandSample(context).ifPresent(staging.get()::stage);
+              return null;
+            }
+            return context;
+          });
+    }
+
+    private Optional<StackTrace> takeOnDemandSample(SamplingContext context) {
+      long currentSampleTime = System.nanoTime();
+      // When the context is locked, the periodic sampling thread is actively reporting
+      // a sample. No need to report both so skip the on-demand sample.
+      if (context.lock.tryLock()) {
+        try {
+          ThreadInfo threadInfo = collector.getThreadInfo(context.thread.getId());
+          SpanContext spanContext = retrieveActiveSpan(context.thread);
+          return toStackTrace(threadInfo, context, spanContext.getSpanId(), currentSampleTime);
+        } finally {
+          context.lock.unlock();
+        }
+      }
+      return Optional.empty();
+    }
+
+    void shutdown() {
+      shutdownLatch.countDown();
+    }
+
+    @Override
+    public void run() {
+      try {
+        while (!Thread.currentThread().isInterrupted()) {
+          boolean shutdown = shutdownLatch.await(delay.toNanos(), TimeUnit.NANOSECONDS);
+          if (shutdown) {
+            return;
+          }
+          sample(traceSamplingContexts.values());
+        }
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+      }
+    }
+
+    private void sample(Collection<SamplingContext> contexts) {
+      if (contexts.isEmpty()) {
+        return;
+      }
+
+      Map<Long, SamplingContext> threadContexts =
+          contexts.stream().collect(Collectors.toMap(c -> c.thread.getId(), context -> context));
+      long currentSampleTime = System.nanoTime();
+      try {
+        ThreadInfo[] threadInfos = collector.getThreadInfo(threadContexts.keySet());
+        List<StackTrace> stackTraces =
+            toStackTraces(threadInfos, threadContexts, currentSampleTime);
+        staging.get().stage(stackTraces);
+      } catch (Exception e) {
+        logger.info("Unexpected error during callstack sampling");
+      }
+    }
+
+    private List<StackTrace> toStackTraces(
+        ThreadInfo[] threadInfos, Map<Long, SamplingContext> contexts, long currentSampleTime) {
+      List<StackTrace> stackTraces = new ArrayList<>(threadInfos.length);
+      for (ThreadInfo threadInfo : threadInfos) {
+        SamplingContext context = contexts.get(threadInfo.getThreadId());
+        // When the context is locked and on demand sample is being taken. No need to report
```

Comment:
should be `an` instead of `and`?
instead of `skip the on-demand sample` should be `skip the periodic sample`?